daemon/src/maker_inc_connections.rs

```diff
@@ -441,6 +377,119 @@ impl Actor {
     }
 }
 
+/// Sets up a new connection on the given stream including all relevant background tasks like
+/// heartbeat and reading messages from the stream.
+async fn setup_new_connection(
+    this: Address<Actor>,
+    noise_priv_key: x25519_dalek::StaticSecret,
+    heartbeat_interval: Duration,
+    stream: TcpStream,
+    address: SocketAddr,
+) {
+    let (mut read, write, taker_id) = match upgrade(stream, noise_priv_key).await {
+        Ok((read, write, identity)) => (read, write, identity),
+        Err(e) => {
+            tracing::warn!(address = %address, "Failed to upgrade incoming connection: {:#}", e);
+            return;
+        }
+    };
+
+    let mut tasks = Tasks::default();
+    tasks.add({
+        let this = this.clone();
+
+        async move {
+            while let Ok(Some(msg)) = read.try_next().await {
+                let res = this.send(FromTaker { taker_id, msg }).await;
+
+                if res.is_err() {
+                    break;
+                }
+            }
+
+            let _ = this.send(ReadFail(taker_id)).await;
+        }
+    });
+    tasks.add({
+        let this = this.clone();
+
+        async move {
+            while let Ok(()) = this.send(SendHeartbeat(taker_id)).await {
+                tokio::time::sleep(heartbeat_interval).await;
+            }
+        }
+    });
+
+    let _ = this
+        .send(ConnectionReady(Connection {
+            _tasks: tasks,
+            taker: taker_id,
+            write,
+        }))
+        .await;
+}
+
+/// Upgrades a TCP stream to an encrypted transport, checking the network version in the process.
+///
+/// Both IO operations, upgrading to noise and checking the version are gated by a timeout.
+async fn upgrade(
+    mut stream: TcpStream,
+    noise_priv_key: x25519_dalek::StaticSecret,
+) -> Result<(
+    wire::Read<wire::TakerToMaker, wire::MakerToTaker>,
+    wire::Write<wire::TakerToMaker, wire::MakerToTaker>,
+    Identity,
+)> {
+    let taker_address = stream.peer_addr().context("Failed to get peer address")?;
+
+    tracing::info!(%taker_address, "Upgrade new connection");
+
+    let transport_state = noise::responder_handshake(&mut stream, &noise_priv_key)
+        .timeout(Duration::from_secs(20))
```

Comment:
So, essentially, this is what we discussed earlier when we came to the conclusion that the handshake might block in some scenarios - Correct?

Reply:
The taker only retries after failed attempts? If we are within this timeout, the handshake is still going on so the taker does not know about the outcome yet?

20 seconds should be plenty.